sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1333,6 +1333,20 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     }
   }
 
+  /**
+   * Create a Substring/Substr expression.
+   */
+  override def visitSubstring(ctx: SubstringContext): Expression = withOrigin(ctx) {
+    val field = ctx.field
+    val pos = ctx.pos
+    val len = ctx.len
+    if (len != null) {
+      Substring(expression(field), expression(pos), expression(len))
+    } else {
+      new Substring(expression(field), expression(pos))
```

Comment:
Why not also use the case class factory as above? It can be good to add a default value to the third arg of the case class and then use the case class factory and get rid of the verbose constructor in that class

Reply:
Add a default value for the case class and delete the verbose constructor may affect `Right` expression. Do we really to do this change in this PR?

Reply:
Whoops, I haven't thought it can be that tricky. Then I don't have a strong opinion about it. cc @dongjoon-hyun

Reply:
Let us keep it unchanged.

Reply:
+1